### DISCUSSION TITLE
Değişken İsimlerine Erişmek İstiyorum Fakat Erişemiyorum

### DISCUSSION THREAD

**Discussion by bgebes** [19/classroom#48] [Ruby] [answered]
### Merhaba Hocam, 

Oluşturmuş olduğum `nilCheck?` fonksiyonuna gönderilen argümanların tek tek nil olup olmadığını kontrol eden bir işlev kodladım. Fakat, `storage` Hash'inin `message` property'sinde nil olan değişkenlerin isimlerini göstermeye çalıştığımda başarısızlıkla karşılaştım. Yapmak istediğim 6.satırda anlık incelemekte olduğum argümanın ismini `fails` dizisine kaydetmekti.

### 1. `local_variables` nesnesi: İçerisinde bulunduğu blokta oluşturulmuş değişkenlerin isimlerini listeler. Fakat, parametresine gönderilmiş olduğu fonksiyonda isimleri yeni isimlerle görür. Burada da göreceğiniz gibi `c` değişkeni olarak görmesi gerekirken `arg` değişkeni olarak görüyor.
```rb
args.each do |arg|
 if(arg.class == NilClass)
    puts local_variable #=> arg, args, storage
    storage[:fails] << arg
    storage[:status] = true
 end
end
```
### 2. `Hash-Key` yöntemi: Bu yöntemde ele almakta olduğum argümanı yeni oluşturduğum Hash'e key olarak (value'sini rastgele birşey) atayıp daha sonra bu Hash'in `Keys` metodundan yararlanarak değişken ismine ulaşmaya çalıştım. Fakat, burada `arg` direk değerini döndürdüğü için herhangi bir sonuçla karşılaşamadım.
```rb
args.each do |arg|
  if(arg.class == NilClass)
    hashing = Hash[arg => 12] 
    storage[:fails] << hashing.keys[0] #=> Konsolda herhangi bir gösterimi yok muhtamelen karşılığı nil olduğu için.
    storage[:status] = true
  end
end
```

<details>
  <summary><h3>Kaynak kod için tıklayabilirsiniz</h3></summary>
  <img src="https://user-images.githubusercontent.com/72809874/168925199-e02507bc-d4d1-4831-8d2f-e558bc5e2d08.png" alt="Kaynak kod" />
</details>

**Comment by roktas** [comment]
Tam olarak ne amaçla böyle bir şeye ihtiyaç duyduğunu bilmediğimden sadece sunduğun koda bakarak cevap vereyim (genelde en doğru cevap niyeti öğrenip ona göre bir çözümleme yapmayı gerektirir).

- `check?` metoduna değişken argüman listesi (`*args` ) gönderiyorsun.  Bu haliyle `a`, `b`, `c`, `d` isimlerini zaten tanımaz, metodun tek gördüğü `args` dizisi.  Amacın ayrı ayrı argümanların ismini almaksa `*args` yerine açık argümanları iletmen gerekir (ör. `def check?(a, b, c, d)`)

- `nil` denetimi yapmanın kanonik yolu `arg.class == NilClass` ifadesi değil, basitçe `arg.nil?`

- Yaptığı iş göz önüne alındığında `check?` uygun bir metod adı değil.  Adı `?` ile biten metodların mantıksal değer dönmesi beklenir, halbuki sende bir hash dönüyor.  Sondaki `?`'nı kaldırmalısın.

**Reply by bgebes** [reply]
Amacım, geliştiricilerin tanımladıkları fonksiyonlar içerisinde işlemler gerçekleşmeden önce fonksiyonun aldığı parametrelerin uygun olup olmadığını kontrol eden bir gem geliştirmek. Bu fonksiyonu örnek olarak kullanma şeklimi şöyle sunabilirim:

```rb
def selamla(insan1, insan2)
 storage = check? insan1, insan2
 raise ArgumentError, storage[:message] if storage[:status]
 
 ...
end
```

### Önerilerinize Cevaplarım
1. Yukarıda gösterdiğim örnek kullanım şeklinde de olduğu gibi bu check fonksiyonunu `farklı farklı fonksiyonlar` içerisinde çağırmak istiyorum. Bu yüzden check fonksiyonunun argümanlarıyla `bağımsız olması` gerektiğini düşünüyorum. Yanlış düşünüyorsam düzeltin lütfen.
2. `nil` denetimini dediğiniz yöntemi kullanarak değiştireceğim teşekkür ederim.
3. `?` ibaresini dediğiniz şekilde kaldıracağım teşekkür ederim.

**Comment by roktas** [comment, marked as answer]
Bir an için bunu yapabileceğini varsaysak bile argüman denetimini çalışma zamanında her metot çağrısında yapman maliyetli bir işlem, bu haliyle pratikte kullanıcı bulman zor buna.  Testler bu tür durumları yakalamak için yazılıyor.  Başlangıç düzeyinde bu tür "meta" gem'ler üzerinde çalışmak yerine daha somut problemlere yoğunlaşmanı öneririm.

**Reply by bgebes** [reply]
Haklısınız hocam, çok teşekkür ederim.
